### PR TITLE
Another ~33% increase in single-thread performance for IlluminaBasecallsToFastq

### PR DESCRIPTION
**Pull request by tfenne**
Two major changes herein:
1. Have `BclReader` read a batch of reads at a time from each cycle's stream and construct multiple `BclData`s at the same time
2. Use a lookup to decode the `byte`s coming out of BCLs into bases and quals

Note: we probably want to do similar things for CbclReader if it's also reading in a similar way.
Also: precomputing lookups for binned quality scores is trickier but probably something we should do too

### CODE REVIEW

File: src/main/java/picard/illumina/parser/readers/BaseBclReader.java [hunk 1]
```diff
@@ -16,16 +16,32 @@
 public class BaseBclReader {
     private static final byte BASE_MASK = 0x0003;
     private static final byte[] BASE_LOOKUP = new byte[]{'A', 'C', 'G', 'T'};
-    private static final byte NO_CALL_BASE = (byte) '.';
+    public static final byte NO_CALL_BASE = (byte) 'N';
```
Comment (jacarey) [inline src/main/java/picard/illumina/parser/readers/BaseBclReader.java:19]: I think this will break basecalls to sam.

**Review by jacarey — CHANGES_REQUESTED** [review]
Aside from the NO_CALL_BASE issue for basecalls to sam it LGTM. Perhaps the converter builder needs a method for setting the NO_CALL_BASE?

File: src/main/java/picard/illumina/parser/readers/BaseBclReader.java (hunk 1 above)
Comment (tfenne) [reply src/main/java/picard/illumina/parser/readers/BaseBclReader.java:19]: Ugh, it does indeed break the tests.  Though it seems odd to me that `IlluminaBasecallsToSam` outputs periods in the reads.  I'm tempted to just update the test cases to Ns.  Thoughts?

Comment (tfenne) [reply src/main/java/picard/illumina/parser/readers/BaseBclReader.java:19]: And round-tripping to BAM and back converts those to Ns.

Comment (jacarey) [reply src/main/java/picard/illumina/parser/readers/BaseBclReader.java:19]: I think that is fine. I'm not sure if there are any additional picard tests that may break in other places (private). 

I'd also worry a bit about anyone that has written similar type integration tests using basecalls to sam and doing direct comparisons

**Review by jacarey — APPROVED** [review]

**Pull request merged**